README.md

Rust to Web assembly experiment
-------------------------------

Drawing a Mandelbrot fractal in the browser,
importing a Rust function compiled to WebAssembly.
Compare speed with raw JS.

See /docs/ to start the project.

### Startup

    source ../emsdk-portable/emsdk_env.sh
    source $HOME/.cargo/env

### Compile Rust files to Web Assembly

    rustc --target=wasm32-unknown-emscripten src/mandelbrot.rs -o weba/mandelbrot.html

The build goes to /weba/. It produces a ".wasm", a ".js" and a ".html" file.

* The .wasm is the WebAssembly binary file.
* The .html is ready to load in a browser with a built-in console.
* The .js is the file to load in a `<script>` tag in order to use wasm functions in JS.

### Run in a web page

- Go to /weba/
- Run a web server at localhost:8000:

      ```
      python -m SimpleHTTPServer
      ```
      
- Make sure web assembly is enabled in the browser
    - Chrome: `chrome://flags` -> "WebAssembly" stuff  
    - Firefox: `about:config` -> "wasm" stuff 
- Go to [localhost:8000/index.html](localhost:8000/index.html)
  in a browser.



history.txt
  525  curl https://sh.rustup.rs -sSf | sh
  529  rustup
  530  vim .profile 
  531  source .profile 
  532  rustup install stable
  533  rustup default stable
  534  rustup target add wasm32-unknown-emscripten
  536  cd Workspace/
  538  mkdir wasm
  539  cd wasm/
  543  brew install wget
  544  wget https://s3.amazonaws.com/mozilla-games/emscripten/releases/emsdk-portable.tar.gz
  546  tar -xvzf emsdk-portable.tar.gz 
  548  rm emsdk-portable.tar.gz
  551  cd emsdk-portable/
  552  source ./emsdk_env.sh
  558  emcc -v   # fail
  563  brew install cmake
  564  emsdk update
  565  emsdk install sdk-incoming-64bit
  566  emsdk activate sdk-incoming-64bit
  570  source $HOME/.cargo/env
  567  emcc -v
  572  rustup component add rust-src
  573  rustup component add rust-docs
  574  rustup update
  575  rustup toolchain install nightly
  588  rustup run nightly cargo install clippy
  577  cd ../
  580  cargo init --bin mytest && cd mytest
  589  cargo install racer
  587  vim src/main.rs 
  594  rustc --target=wasm32-unknown-emscripten src/main.rs -o hello.html
  596  cd ..
  598  source ./emsdk-portable/emsdk_env.sh
  599  emcc -v
  600  cd mytest
  601  rustc --target=wasm32-unknown-emscripten src/hello.rs -o weba/hello.html


weba/index.js

// Ref: http://progur.com/2017/02/create-mandelbrot-fractal-javascript.html


(function() {

    var maxIterations = 100;
    var magnificationFactor = 300;  // 2900
    var panX = 1.5;  // 0.6
    var panY = 1;  // 0.7
    var SIZE = 500;

    Module['onRuntimeInitialized'] = onRuntimeInitialized;
    function onRuntimeInitialized() {
        jsFractal();
        webaFractal();
    }

    /**
     * ----   WASM   ----
     */
    function webaFractal() {
        var canvas = newCanvas();
        var ctx = canvas.getContext("2d");
        var timeWeba = 0;
        var iterations = 0;

        for (var x = 0; x < canvas.width; x++) {
            for (var y = 0; y < canvas.height; y++) {

                var xx = x / magnificationFactor - panX;
                var yy = y / magnificationFactor - panY;

                /* WASM !!! */
                var t0 = performance.now();
                var belongsToSet = _belongs_to_mandelbrot_set(xx, yy, maxIterations);
                var t1 = performance.now();
                timeWeba += t1 - t0;
                iterations += 1;

                if (x % 100 < 1e-8 && y % 100 < 1e-8) {
                    console.log("Wasm:", t1 - t0, timeWeba);
                }

                drawPoint(ctx, x, y, belongsToSet);
            }
        }
        console.log("Wasm time:", timeWeba);
    }

    /**
     * ----   JS   ----
     */
    function jsFractal() {
        var canvas = newCanvas();
        var ctx = canvas.getContext("2d");
        var timeJS = 0;
        var iterations = 0;

        // warmup
        belongsToMandelbrotSet(0, 0, maxIterations);

        for (var x = 0; x < canvas.width; x++) {
            for (var y = 0; y < canvas.height; y++) {

                var xx = x / magnificationFactor - panX;
                var yy = y / magnificationFactor - panY;

                /* JS */
                var t0 = performance.now();
                var belongsToSet = belongsToMandelbrotSet(xx, yy, maxIterations);
                var t1 = performance.now();
                timeJS += t1 - t0;
                iterations += 1;

                if (x % 100 < 1e-8 && y % 100 < 1e-8) {
                    console.log("JS:", t1 - t0, timeJS);
                }

                drawPoint(ctx, x, y, belongsToSet);
            }
        }
        console.log("JS time:", timeJS);
    }

    /** JS version */
    function belongsToMandelbrotSet(x, y, maxIterations) {
        var realComponentOfResult = x;
        var imaginaryComponentOfResult = y;
        for (var i = 0; i <= maxIterations; i++) {
            var tempRealComponent = realComponentOfResult * realComponentOfResult
                - imaginaryComponentOfResult * imaginaryComponentOfResult
                + x;
            var tempImaginaryComponent = 2 * realComponentOfResult * imaginaryComponentOfResult
                + y;
            realComponentOfResult = tempRealComponent;
            imaginaryComponentOfResult = tempImaginaryComponent;

            // Color: return as a percentage
            if (realComponentOfResult * imaginaryComponentOfResult > 5) {
                return (i / maxIterations * 100);
            }
        }
        return 0;   // Return zero if in set
    }

    function drawPoint(ctx, x, y, belongsToSet) {
        if (belongsToSet) {
            if (belongsToSet === 0) {
                ctx.fillStyle = '#000';
                ctx.fillRect(x,y, 1,1); // Draw a black pixel
            } else {
                ctx.fillStyle = 'hsl(0, 100%, ' + belongsToSet + '%)';
                ctx.fillRect(x,y, 1,1); // Draw a colorful pixel
            }
        }
    }

    function newCanvas() {
        var canvas = document.createElement("canvas");
        canvas.width = SIZE;
        canvas.height = SIZE;
        document.body.appendChild(canvas);
        return canvas;
    }

})();



src/mandelbrot.rs

// Functions that you wish to access from Javascript must be marked as no_mangle
#[no_mangle]
pub fn belongs_to_mandelbrot_set(x: f64, y: f64, max_iterations: u64) -> f64 {
    // println!("{}", format!("!!! {} {} {}", x, y, max_iterations));

    let mut real_component_of_result: f64 = x;
    let mut imaginary_component_of_result: f64 = y;

    for i in 0..max_iterations {
        let temp_real_component: f64 = real_component_of_result * real_component_of_result
            - imaginary_component_of_result * imaginary_component_of_result
            + x;
        let temp_imaginary_component: f64 = 2.0 * real_component_of_result * imaginary_component_of_result
            + y;
        real_component_of_result = temp_real_component;
        imaginary_component_of_result = temp_imaginary_component;

        // Color: return a percentage
        if real_component_of_result * imaginary_component_of_result > 5.0 {
            return i as f64 / max_iterations as f64 * 100.0;
        }
    }
    return 0.0;   // Return zero if in set
}

fn main() {
    println!("mais.rs: My fractal application in Rust");
}



weba/mandelbrotJS.js

// Ref: http://progur.com/2017/02/create-mandelbrot-fractal-javascript.html


(function() {

    var maxIterations = 100;
    var magnificationFactor = 300;  // 2900
    var panX = 1.5;  // 0.6
    var panY = 1;  // 0.7
    var SIZE = 500;

    Module['onRuntimeInitialized'] = onRuntimeInitialized;
    function onRuntimeInitialized() {
        jsFractal();
        webaFractal();
    }

    /**
     * ----   WASM   ----
     */
    function webaFractal() {
        var canvas = newCanvas();
        var ctx = canvas.getContext("2d");
        var timeWeba = 0;
        var iterations = 0;

        for (var x = 0; x < canvas.width; x++) {
            for (var y = 0; y < canvas.height; y++) {

                var xx = x / magnificationFactor - panX;
                var yy = y / magnificationFactor - panY;

                /* WASM !!! */
                var t0 = performance.now();
                var belongsToSet = _belongs_to_mandelbrot_set(xx, yy, maxIterations);
                var t1 = performance.now();
                timeWeba += t1 - t0;
                iterations += 1;

                if (x % 100 < 1e-8 && y % 100 < 1e-8) {
                    console.log("Wasm:", t1 - t0, timeWeba);
                }

                drawPoint(ctx, x, y, belongsToSet);
            }
        }
        console.log("Wasm time:", timeWeba, "Iterations:", iterations);
    }

    /**
     * ----   JS   ----
     */
    function jsFractal() {
        var canvas = newCanvas();
        var ctx = canvas.getContext("2d");
        var timeJS = 0;
        var iterations = 0;

        // warmup
        belongsToMandelbrotSet(0, 0, maxIterations);

        for (var x = 0; x < canvas.width; x++) {
            for (var y = 0; y < canvas.height; y++) {

                var xx = x / magnificationFactor - panX;
                var yy = y / magnificationFactor - panY;

                /* JS */
                var t0 = performance.now();
                var belongsToSet = belongsToMandelbrotSet(xx, yy, maxIterations);
                var t1 = performance.now();
                timeJS += t1 - t0;
                iterations += 1;

                if (x % 100 < 1e-8 && y % 100 < 1e-8) {
                    console.log("JS:", t1 - t0, timeJS);
                }

                drawPoint(ctx, x, y, belongsToSet);
            }
        }
        console.log("JS time:", timeJS, "Iterations:", iterations);
    }

    /** JS version */
    function belongsToMandelbrotSet(x, y, maxIterations) {
        var realComponentOfResult = x;
        var imaginaryComponentOfResult = y;
        for (var i = 0; i <= maxIterations; i++) {
            var tempRealComponent = realComponentOfResult * realComponentOfResult
                - imaginaryComponentOfResult * imaginaryComponentOfResult
                + x;
            var tempImaginaryComponent = 2 * realComponentOfResult * imaginaryComponentOfResult
                + y;
            realComponentOfResult = tempRealComponent;
            imaginaryComponentOfResult = tempImaginaryComponent;

            // Color: return as a percentage
            if (realComponentOfResult * imaginaryComponentOfResult > 5) {
                return (i / maxIterations * 100);
            }
        }
        return 0;   // Return zero if in set
    }

    function drawPoint(ctx, x, y, belongsToSet) {
        if (belongsToSet) {
            if (belongsToSet === 0) {
                ctx.fillStyle = '#000';
                ctx.fillRect(x,y, 1,1); // Draw a black pixel
            } else {
                ctx.fillStyle = 'hsl(0, 100%, ' + belongsToSet + '%)';
                ctx.fillRect(x,y, 1,1); // Draw a colorful pixel
            }
        }
    }

    function newCanvas() {
        var canvas = document.createElement("canvas");
        canvas.width = SIZE;
        canvas.height = SIZE;
        document.body.appendChild(canvas);
        return canvas;
    }

})();

